llvm/lib/Target/AMDGPU/SIPreEmitPeephole.cpp

```diff
@@ -304,11 +299,67 @@ bool SIPreEmitPeephole::getBlockDestinations(
   return true;
 }
 
+namespace {
+class BranchWeightCostModel {
+  const SIInstrInfo &TII;
+  const TargetSchedModel &SchedModel;
+  BranchProbability BranchProb;
+  uint64_t BranchCost;
+  uint64_t ThenCyclesCost = 0;
+
+public:
+  BranchWeightCostModel(const SIInstrInfo &TII, const MachineInstr &Branch,
+                        const MachineBasicBlock &Succ)
+      : TII(TII), SchedModel(TII.getSchedModel()) {
+    assert(SchedModel.hasInstrSchedModelOrItineraries());
+
+    const MachineBasicBlock &Head = *Branch.getParent();
+    const auto *FromIt = find(Head.successors(), &Succ);
+    assert(FromIt != Head.succ_end());
+
+    BranchProb = Head.getSuccProbability(FromIt);
+    if (BranchProb.isUnknown())
+      return;
+
+    BranchCost = SchedModel.computeInstrLatency(&Branch, false);
+  }
+
+  bool isUnknown() const { return BranchProb.isUnknown(); }
+
+  bool isProfitable(const MachineInstr &MI) {
+    assert(!isUnknown());
+
+    if (TII.isWaitcnt(MI.getOpcode()))
+      return false;
+
+    ThenCyclesCost += SchedModel.computeInstrLatency(&MI, false);
```

Comment:
I don't think adding up the latency of all the instructions is the right metric. E.g (on GFX10+) typical VALU instructions can be issued one per cycle if there are no dependencies, but they have 5 cycle latency before you can use the result.

Or have I misunderstood what computeInstrLatency does, vs computeOperandLatency?

Reply:
Ideally we would have something that uses the scheduling model to estimate the cost of a block of code, based on instruction issue and on latency if (and only if) there are dependent instructions. I don't know if anything like that exists yet.

Reply:
To compare the results when using `MachineTraceMetrics` I've created https://github.com/llvm/llvm-project/pull/111117 .